src/main.rs

```diff
@@ -540,3 +547,180 @@ mod provision {
+            leaf: 0x8000001f,
+            func: |res| (res.eax & 0x1 != 0, None),
+        },
+        CpuId {
+            name: "Secure Encrypted Virtualization (SEV)",
+            leaf: 0x8000001f,
+            func: |res| (res.eax & (0x1 << 1) != 0, None),
+        },
+        CpuId {
+            name: "Page flush MSR",
+            leaf: 0x8000001f,
+            func: |res| (res.eax & (0x1 << 2) != 0, None),
+        },
+        CpuId {
+            name: "VM Permission Levels (VMPL)",
+            leaf: 0x8000001f,
+            func: |res| (res.eax & (0x1 << 5) != 0, None),
+        },
+        CpuId {
+            name: "Physical address bit reduction",
+            leaf: 0x8000001f,
+            func: |res| {
+                let field = res.ebx & 0b1111_1100_0000 >> 6;
+                (true, Some(format!("{}", field)))
+            },
+        },
+        CpuId {
+            name: "C-bit location in page table entry",
+            leaf: 0x8000001f,
+            func: |res| {
+                let field = res.ebx & 0b01_1111;
+                (true, Some(format!("{}", field)))
+            },
+        },
+        CpuId {
+            name: "Number of encrypted guests supported simultaneously",
+            leaf: 0x8000001f,
+            func: |res| (true, Some(format!("{}", res.ecx))),
+        },
+        CpuId {
+            name: "Minimum ASID value for SEV-enabled, SEV-ES disabled guest",
+            leaf: 0x8000001f,
+            func: |res| (true, Some(format!("{}", res.edx))),
+        },
+    ];
+
+    // SEV-ES specific CPUIDs.
+    const SEV_ES_CPUIDS: &[CpuId] = &[CpuId {
+        name: "SEV Encrypted State (SEV-ES)",
+        leaf: 0x8000001f,
+        func: |res| (res.eax & (0x1 << 3) != 0, None),
+    }];
+
+    pub fn cmd(gen: SevGeneration, quiet: bool) -> Result<()> {
+        let mut all_tests_passed = true;
+        let results = ok_run_tests(gen);
+        for tr in results {
+            if !tr.success {
+                all_tests_passed = false;
+            }
+            if !quiet {
+                println!("{}", tr.msg);
+            }
+        }
+
+        if all_tests_passed {
+            Ok(())
+        } else {
+            Err(error::Context::new(
+                "One or more tests in sevctl-ok reported a failure",
+                Box::<Error>::new(ErrorKind::InvalidData.into()),
+            ))
+        }
+    }
+
+    fn ok_run_tests(gen: SevGeneration) -> Vec<OkTestResult> {
+        let mut data = Vec::new();
+
+        data.append(&mut has_processor_support(gen));
+
+        data
+    }
+
+    fn has_processor_support(gen: SevGeneration) -> Vec<OkTestResult> {
+        let mut c_vec = vec![SEV_CPUIDS];
+        if let SevGeneration::Es = gen {
+            c_vec.push(SEV_ES_CPUIDS);
+        }
+
+        let mut results = vec![];
+        for id in c_vec.iter_mut().flat_map(|r| r.iter()) {
+            results.push(id.execute());
+        }
+
+        results
+    }
```

Comment:
This function is called `has_processor_support` but it returns a list of test results, some of which may have failed. The function itself doesn't directly answer the question as to whether the processor supports the given generation or not, but hands the caller a bunch of results to deduce that for themselves.